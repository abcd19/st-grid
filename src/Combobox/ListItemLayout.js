
import React from 'react';
import ReactDOM from 'react-dom';

/**
 * @class
 */
class ListItemLayout extends React.Component{
    
    /**
     * @constructor
     * @param {type} data
     */
    constructor(props)
    {
      super(props);
      this.onMouseOverHandle = this.onMouseOverHandle.bind(this);
      this.onMouseOutHandle = this.onMouseOutHandle.bind(this);
      this.onClickHandle = this.onClickHandle.bind(this);
      this.state = {
        mouseOver: false
      }
      this.div = React.createRef();
    };

    markMouseOver(on)
    {
      if(on === true)
      {
        this.setState({mouseOver: true})
      }else{
        this.setState({mouseOver: false})
      }
    }

    onMouseOverHandle()
    {
      this.markMouseOver(true)
    }
    
    onMouseOutHandle()
    {
      this.markMouseOver(false)
    }

    onClickHandle()
    {
      this.props.onClick(this.props.val);
    }

    componentDidMount()
    {
      if(this.props.setScroll == true)
      {
        this.div.current.scrollIntoView({ behavior: 'auto', block: "center" })
      }
    }



    render(){
      
      let style ={
        borderTop: '1px lightgray solid',
        height: '18px',
        whiteSpace: 'nowrap',
        lineHeight: 'normal',
        fontKerning: 'auto',
        fontFamily: '"Helvetica Neue", Helvetica, Arial, sans-serif',
        fontSize: '12px',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        paddingTop: '2px',
        paddingBottom: '2px',
        boxSizing: 'content-box',
      }
 
      if(this.state.mouseOver == true)
      {
        style.background = '#FEFFBF'
      }else{

        if(this.props.isSelected == true)
        {
          style.background = '#EDF5FC';
        }else{
          style.background = 'white';
        }
        
      }

      return(
        <div ref={ this.div } style={style} onClick={ this.onClickHandle } onMouseOver={ this.onMouseOverHandle } onMouseOut={ this.onMouseOutHandle }> { this.props.val.display }</div>
      );
    }
    
};
export {ListItemLayout}