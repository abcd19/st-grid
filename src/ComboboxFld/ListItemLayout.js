
import React, {useEffect, useState, useRef} from 'react';

/**
 * Элемент списка
 */
export const ListItemLayout = (props) =>
{
  let [mouseOver, setMouseOver] = useState(false);

  let div = useRef(null);

  useEffect(()=>{
    if(props.setScroll == true)
    {
      div.current.scrollIntoView({ behavior: 'auto', block: "center" })
    }
  });

  let style ={
    borderTop: '1px lightgray solid',
    minHeight: '18px',
    //wordWrap: 'break-word',
    whiteSpace: 'normal',
    lineHeight: 'normal',
    fontKerning: 'auto',
    fontFamily: '"Helvetica Neue", Helvetica, Arial, sans-serif',
    fontSize: '12px',
    width: props.itemsListWidthPix,
    //overflow: 'hidden',
    //textOverflow: 'ellipsis',
    paddingTop: '2px',
    paddingBottom: '2px',
    boxSizing: 'content-box',
  }

  if(mouseOver == true)
  {
    style.background = '#FEFFBF'
  }else{

    if(props.isSelected == true)
    {
      style.background = '#EDF5FC';
    }else{
      style.background = 'white';
    }
    
  }

  let onClick = (e) =>
  {
    props.onClick(props.val);
    e.stopPropagation();
  }

  return(
    <div 
      title = { props.val.display }
      ref= {div}
      style={style} 
      onClick={ onClick } 
      onMouseDown={(e) => e.stopPropagation()} 
      onMouseUp = {(e) => e.stopPropagation()}
      onMouseOver={ () => setMouseOver(true) } 
      onMouseOut={ () => setMouseOver(false) }> { props.val.display }</div>
  );
  
}