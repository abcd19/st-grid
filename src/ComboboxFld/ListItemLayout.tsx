
import React, { useEffect, useState, useRef, SyntheticEvent } from 'react';

export type typeVal = {
  raw?: string | number | boolean;
  display?: string | number | boolean;
};

export interface IListItemLayoutProps {
  setScroll?: boolean;
  itemsListWidthPix?: number;
  isSelected?: boolean;
  onClick?: (val: typeVal) => void;
  val?: typeVal;
}

// combobox list element
export const ListItemLayout: React.FC<IListItemLayoutProps> = (props: IListItemLayoutProps) => {
  const [mouseOver, setMouseOver] = useState(false);

  const div = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (props.setScroll == true) {
      if (div && div.current) {
        div.current.scrollIntoView({ behavior: 'auto', block: "center" })
      }
    }
  });

  const style: React.CSSProperties = {
    borderTop: '1px lightgray solid',
    minHeight: '18px',
    //wordWrap: 'break-word',
    whiteSpace: 'normal',
    lineHeight: 'normal',
    fontKerning: 'auto',
    fontFamily: '"Helvetica Neue", Helvetica, Arial, sans-serif',
    fontSize: '12px',
    width: props.itemsListWidthPix,
    //overflow: 'hidden',
    //textOverflow: 'ellipsis',
    paddingTop: '2px',
    paddingBottom: '2px',
    boxSizing: 'content-box',
  }

  if (mouseOver == true) {
    style.background = '#FEFFBF'
  } else {

    if (props.isSelected == true) {
      style.background = '#EDF5FC';
    } else {
      style.background = 'white';
    }

  }

  const onClick = (e: SyntheticEvent<HTMLDivElement>) => {
    if (typeof props.onClick == 'function' && props.val != undefined) {
     
      props.onClick(props.val);
    }

    e.stopPropagation();
  }

  let display = undefined;

  if(props.val && props.val.display)
  {
    display = String(props.val.display);
  }

  return (
    <div
      title={display}
      ref={div}
      style={style}
      onClick={onClick}
      onMouseDown={(e) => e.stopPropagation()}
      onMouseUp={(e) => e.stopPropagation()}
      onMouseOver={() => setMouseOver(true)}
      onMouseOut={() => setMouseOver(false)}> { display}</div>
  );

}