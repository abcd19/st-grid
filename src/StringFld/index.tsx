export {ImgButtonLayout, IImgButtonLayoutProps} from './ImgButtonLayout';
export {InputLayout, IInputLayoutProps} from './InputLayout';
export {LinearGroupLayout, ILinearGroupLayoutProps} from './LinearGroupLayout';
export {FieldLayoutEdit, IFieldLayoutEditProps} from './FieldLayoutEdit';
export {StringFldLayoutEdit, IStringFldLayoutEditProps, typeStringFldVal, typeStringFldOnChange} from './StringFldLayoutEdit';
export {StringFldCell, StringFldCellProps} from './StringFldCell';

import {StringFldLayoutEdit} from './StringFldLayoutEdit';
export default StringFldLayoutEdit;