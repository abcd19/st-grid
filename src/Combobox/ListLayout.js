import * as ST from './../common'
import React from 'react';
import ReactDOM from 'react-dom';
import {ListItemLayout} from './ListItemLayout'
import {SearchFldLayout} from './SearchFldLayout'
import {calcRank} from './../findWords';

/**
 * @class
 */
class ListLayout extends React.Component{
    
    /**
     * @constructor
     * @param {type} data
     */
    constructor(props)
    {
      super(props);
      this.clickItemHandle = this.clickItemHandle.bind(this);
      this.genDiv = React.createRef();
      this.changeSearchHandle = this.changeSearchHandle.bind(this);
      this.state ={
        //массив с отфильтрованными строками при поиске
        filterdItems: [],
        //флаг фильтр пременен
        isFiltred: false
      }
      this.onWheelHandle = this.onWheelHandle.bind(this);
    };

    clickItemHandle(val)
    {
      if(ST.has(this.props,'handler.change'))
      {
        this.props['handler']['change'](val);
      }
    }

    onWheelHandle(e)
    {
      //при скроллинге листа скроллится только он
      //Если убрать то скроллиться грид      
      e.stopPropagation();
    }

    //фильтрация элементов при поиске
    changeSearchHandle(val)
    {
      //ищем по значениям для отображения
      let ar = [];
      for(var i = 0;  i < this.props.items.length; i++)
      {
        ar[i] = this.props.items[i]['display'];
      }

      //ранжируем массив
      var ranked = calcRank(ar,  val,{
        deleteElementByRankZero: true //удаляем элементы с нулевым рангом
      });

      //создаем новый массив элементов, который удовлетворяет поиску
      let newItems = [];

      //Если ничего не найдено
      /*if(ranked.length === 0)
      {
        //newItems = this.props.items;
        this.setState({filterdItems: [], isFiltred: false});
        return;
      }*/

      for(var i = 0; i < ranked.length; i++)
      {
          var elem = ranked[i]['elemLink'];
          newItems.push(this.props.items[elem]);
      }

      this.setState({filterdItems: newItems, isFiltred: true});
      console.dir(newItems);
    }

    componentDidMount() {
      //расчет позиции для отображения
      let {left, top, width} = this.genDiv.current.parentNode.getBoundingClientRect();
      this.genDiv.current.style.width = width+'px';
     
      //если один из парентов имеет позицию fixed или absolute то позицию нужно считать от него
      if(this.genDiv.current.offsetParent && this.genDiv.current.offsetParent.nodeName != 'BODY')
      {
        //console.dir(this.genDiv.current.offsetParent)
        var offsetParent = this.genDiv.current.offsetParent.getBoundingClientRect();
        
        left =  left - offsetParent.left;
        top =  top - offsetParent.top;
      }

      //console.dir(this.genDiv.current.offsetParent.getBoundingClientRect());
      function getBodyScrollTop()
      {
        return (document.documentElement && document.documentElement.scrollTop) || 
        (document.body && document.body.scrollTop);
      }

      function getBodyScrollLeft()
      {
        return (document.documentElement && document.documentElement.scrollLeft) || 
        (document.body && document.body.scrollLeft);
      }

      // у потомка берем его отступ и прибавляем размер скролла страницы
      this.genDiv.current.style.left = left + getBodyScrollLeft() + 'px';
      this.genDiv.current.style.top = top + getBodyScrollTop() + 'px';
    }

    render(){
      
      let items = [];

      //Если нет фильтрации
      if(this.state.isFiltred == false)
      {

        for(let i = 0; i < this.props.items.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.props.items[i].raw)
          {
            isSelected = true;
          }

          let newItem = <ListItemLayout setScroll = {isSelected} isSelected= {isSelected} key ={i} val ={ this.props.items[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }
      }else{ //Если есть фильтрация

        for(let i = 0; i < this.state.filterdItems.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.state.filterdItems[i].raw)
          {
            isSelected = true;
          }

          let newItem = <ListItemLayout setScroll = {false}  isSelected= {isSelected} key ={i} val ={ this.state.filterdItems[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }

      }

      let searchBlock;
      //поле поиска показываем только если айтемов больше 12
      if(!this.props.disableSearch && this.props.items.length > 12)
      {
        searchBlock = <tr>
                          <td>
                            <div><SearchFldLayout onChangeDelay = { this.changeSearchHandle } /></div>
                          </td>
                        </tr>;
      }
      
      const itemsVisibleMax = 7;

      const itemHeight = 23;

      let itemsDivHeight = itemsVisibleMax * itemHeight;
      if(items.length < itemsVisibleMax)
      {
        itemsDivHeight = items.length*itemHeight
      }

      return(
        <div id="ListLayout" ref={this.genDiv}  style= {{background: 'white', width: "100%",  boxSizing: 'border-box', zIndex:'100', position: 'absolute', border: '1px solid gray'}}>
          <table style={{borderSpacing: '2px', borderCollapse: 'separate', width: '100%'}}>
            <tbody>
              {searchBlock}
              <tr>
                <td style={{padding: '0px', margin: '0px'}}>
                  <div id="ListLayout-item" style={{height: itemsDivHeight+'px', overflow: 'auto', boxSizing: 'content-box'}} onWheel={this.onWheelHandle}>
                    {items}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
      </div>);
    }
};

export {ListLayout}