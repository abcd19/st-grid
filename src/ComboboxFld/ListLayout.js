import * as ST from '../common'
import React from 'react';
import {ListItemLayout} from './ListItemLayout'
import {SearchFldLayoutEdit} from './SearchFldLayoutEdit'
import {calcRank} from '../findWords';

//ширина списка при котрытии
export const WIDTH = 300;
//ширина элемента списка при открытии
export const LIST_ITEM_WIDTH = WIDTH-5;
//максимальная высота списка
export const MAXHEIGHT = 250;

/**
 * Список
 */
export class ListLayout extends React.Component{
    
    /**
     * @constructor
     * @param {type} data
     */
    constructor(props)
    {
      super(props);
      this.clickItemHandle = this.clickItemHandle.bind(this);
      this.genDiv = React.createRef();

      this.changeSearchHandle = this.changeSearchHandle.bind(this);
      this.state ={
        //массив с отфильтрованными строками при поиске
        filterdItems: [],
        //флаг фильтр пременен
        isFiltred: false
      }
      //this.onWheelHandle = this.onWheelHandle.bind(this);
    };

    clickItemHandle(val)
    {
      if(ST.has(this.props,'handler.change'))
      {
        this.props['handler']['change'](val);
      }
    }

    /*onWheelHandle(e)
    {
      //при скроллинге листа скроллится только он
      //Если убрать то скроллиться грид      
      e.stopPropagation();
    }*/

    //фильтрация элементов при поиске
    changeSearchHandle(val)
    {
      //ищем по значениям для отображения
      let ar = [];
      for(var i = 0;  i < this.props.items.length; i++)
      {
        ar[i] = this.props.items[i]['display'];
      }

      //ранжируем массив
      var ranked = calcRank(ar,  val,{
        deleteElementByRankZero: true //удаляем элементы с нулевым рангом
      });

      //создаем новый массив элементов, который удовлетворяет поиску
      let newItems = [];

      //Если ничего не найдено
      /*if(ranked.length === 0)
      {
        //newItems = this.props.items;
        this.setState({filterdItems: [], isFiltred: false});
        return;
      }*/

      for(var i = 0; i < ranked.length; i++)
      {
          var elem = ranked[i]['elemLink'];
          newItems.push(this.props.items[elem]);
      }

      this.setState({filterdItems: newItems, isFiltred: true});
      //console.dir(newItems);
    }

    componentDidMount() {
      let {cordBtnLeft, cordBtnTop, listWidthPix} = this.props;
      
      if(ST.isNumber(listWidthPix) == false)
      {
        listWidthPix = WIDTH;
      }

      //если один из парентов имеет позицию fixed или absolute то позицию нужно считать от него
      if(this.genDiv.current.offsetParent && this.genDiv.current.offsetParent.nodeName != 'BODY')
      {
        //console.dir(this.genDiv.current.offsetParent)
        var offsetParent = this.genDiv.current.offsetParent.getBoundingClientRect();
        
        cordBtnLeft =  cordBtnLeft - offsetParent.left;
        cordBtnTop =  cordBtnTop - offsetParent.top;
      }

      //слева отнимаем ширину ячейки таблицы + небольшая виличина для красоты
      let left = cordBtnLeft - listWidthPix + 30;
      //снизу высоту инпута + небольшая виличина для красоты
      let top = cordBtnTop + 30;

      //console.dir(this.genDiv.current.offsetParent.getBoundingClientRect());
      function getBodyScrollTop()
      {
        return (document.documentElement && document.documentElement.scrollTop) || 
        (document.body && document.body.scrollTop);
      }

      function getBodyScrollLeft()
      {
        return (document.documentElement && document.documentElement.scrollLeft) || 
        (document.body && document.body.scrollLeft);
      }

      top = top + getBodyScrollTop(); 
      let r = document.documentElement.clientHeight-Math.ceil(top)-20;
      if(r <= MAXHEIGHT)
      {
        top = top - MAXHEIGHT;
      }

      // у потомка берем его отступ и прибавляем размер скролла страницы
      this.genDiv.current.style.left = left + getBodyScrollLeft() + 'px';
      this.genDiv.current.style.top = top + 'px';

          //чтобы не крутилась страница в chrom дисеблим скролл с флагом  passive: false
      const prevDef = (e) => e.stopPropagation();
      
      if ('onwheel' in document)
      {
        // IE9+, FF17+, Ch31+
        this.genDiv.current.addEventListener("wheel", prevDef, { passive: false });
      } else if ('onmousewheel' in document) {
        // устаревший вариант события
        this.genDiv.current.addEventListener("mousewheel", prevDef, { passive: false });
      } else {
        // Firefox < 17
        this.genDiv.current.addEventListener("MozMousePixelScroll", prevDef, { passive: false });
      }


    }

    render(){
      
      let items = [];
      let {listWidthPix} = this.props;
      
      //ширина элемента списка при открытии
      let itemsListWidthPix;
      if(ST.isNumber(listWidthPix) == false)
      {
        listWidthPix = WIDTH;
        itemsListWidthPix = listWidthPix-5;
      }else{
        itemsListWidthPix = listWidthPix-5;
      }
      
            
      //Если нет фильтрации
      if(this.state.isFiltred == false)
      {

        for(let i = 0; i < this.props.items.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.props.items[i].raw)
          {
            isSelected = true;
          }

          let newItem = <ListItemLayout setScroll = {isSelected} itemsListWidthPix={itemsListWidthPix} isSelected= {isSelected} key ={i} val ={ this.props.items[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }
      }else{ //Если есть фильтрация

        for(let i = 0; i < this.state.filterdItems.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.state.filterdItems[i].raw)
          {
            isSelected = true;
          }
          //при поиске скролим на первый элемент
          let newItem = <ListItemLayout setScroll = {i == 0?true: false}  isSelected= {isSelected} key ={i} val ={ this.state.filterdItems[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }

      }

      let searchBlock;
      //поле поиска показываем только если айтемов больше 12
      if(!this.props.disableSearch && this.props.items.length > 12)
      {
        searchBlock = <tr>
                          <td>
                            <div><SearchFldLayoutEdit onChangeDelay = { this.changeSearchHandle } /></div>
                          </td>
                        </tr>;
      }
      
      return(
        <div id="ListLayout" ref={this.genDiv}  style= {{background: 'white', width: listWidthPix,  boxSizing: 'border-box', zIndex:'100', position: 'absolute', border: '1px solid gray'}}>
          <table style={{borderSpacing: '2px', width: listWidthPix, borderCollapse: 'separate', width: '100%'}}>
            <tbody>
              {searchBlock}
              <tr>
                <td style={{padding: '0px', margin: '0px'}}>
                  <div id="ListLayout-item" style={{maxHeight: MAXHEIGHT, width: itemsListWidthPix, overflow: 'auto', boxSizing: 'content-box'}} /*onWheel={this.onWheelHandle}*/>
                    {items}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
      </div>);
    }
};
