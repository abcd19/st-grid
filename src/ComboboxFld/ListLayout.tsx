import * as ST from '../common'
import React from 'react';
import {ListItemLayout} from './ListItemLayout'
import {SearchFldLayoutEdit} from './SearchFldLayoutEdit'
import {calcRank} from '../findWords';

//ширина списка при котрытии
export const WIDTH = 300;
//ширина элемента списка при открытии
export const LIST_ITEM_WIDTH = WIDTH-5;
//максимальная высота списка
export const MAXHEIGHT = 250;

export type typeComboValue = {
  raw?: string | number | boolean,
  display?: string | number | boolean,
}

export interface IListLayoutProps {
  handler: {
    change: (val?: typeComboValue) => void
  };
  items: typeComboValue[];
  cordBtnLeft: number;
  cordBtnTop: number;
  listWidthPix: number;
  disableSearch?: boolean;
  selectedVal?: string | number | boolean;
}

export interface IListLayoutState {
  isFiltred: boolean;
  filterdItems: typeComboValue[];
}

// combobox list
export class ListLayout extends React.Component<IListLayoutProps, IListLayoutState>{
    
    private genDiv: React.RefObject<HTMLDivElement>;

    /**
     * @constructor
     * @param {type} data
     */
    constructor(props: IListLayoutProps)
    {
      super(props);
      this.clickItemHandle = this.clickItemHandle.bind(this);
      this.genDiv = React.createRef();

      this.changeSearchHandle = this.changeSearchHandle.bind(this);
      this.state ={
        filterdItems: [],
        isFiltred: false
      }
    }

    clickItemHandle(val?: typeComboValue): void
    {
      if(ST.has(this.props,'handler.change'))
      {
        this.props['handler']['change'](val);
      }
    }

    changeSearchHandle(val?: string): void
    {
      const ar = [];
      for(let i = 0;  i < this.props.items.length; i++)
      {
        ar[i] = this.props.items[i]['display'];
      }

      const ranked = calcRank(ar,  val,{
        deleteElementByRankZero: true
      });

      const newItems = [];

      for(let i = 0; i < ranked.length; i++)
      {
          const elem = ranked[i]['elemLink'];
          newItems.push(this.props.items[elem]);
      }

      this.setState({filterdItems: newItems, isFiltred: true});
    }

    componentDidMount(): void {
      let {cordBtnLeft, cordBtnTop, listWidthPix} = this.props;
      
      if(ST.isNumber(listWidthPix) == false)
      {
        listWidthPix = WIDTH;
      }

      if(this.genDiv && this.genDiv.current 
        && this.genDiv.current.offsetParent && this.genDiv.current.offsetParent.nodeName != 'BODY')
      {
        const offsetParent = this.genDiv.current.offsetParent.getBoundingClientRect();
        
        cordBtnLeft =  cordBtnLeft - offsetParent.left;
        cordBtnTop =  cordBtnTop - offsetParent.top;
      }

      const left = cordBtnLeft - listWidthPix + 30;
      let top = cordBtnTop + 30;
      function getBodyScrollTop()
      {
        return (document.documentElement && document.documentElement.scrollTop) || 
        (document.body && document.body.scrollTop);
      }

      function getBodyScrollLeft()
      {
        return (document.documentElement && document.documentElement.scrollLeft) || 
        (document.body && document.body.scrollLeft);
      }

      top = top + getBodyScrollTop(); 
      const r = document.documentElement.clientHeight-Math.ceil(top)-20;
      if(r <= MAXHEIGHT)
      {
        top = top - MAXHEIGHT;
      }

      if(this.genDiv && this.genDiv.current)
      {
        this.genDiv.current.style.left = left + getBodyScrollLeft() + 'px';
        this.genDiv.current.style.top = top + 'px';
      }
      
      const prevDef = (e: Event) => { e.stopPropagation() };

      if(this.genDiv && this.genDiv.current)
      {
        if ('onwheel' in document)
        {
          // IE9+, FF17+, Ch31+
          this.genDiv.current.addEventListener("wheel", prevDef, { passive: false });
        } else if ('onmousewheel' in document) {
          this.genDiv.current.addEventListener("mousewheel", prevDef, { passive: false});
        } else {
          // Firefox < 17
          this.genDiv.current.addEventListener("MozMousePixelScroll", prevDef, { passive: false});
        }
      }

    }

    render(): React.ReactElement {
      
      const items = [];
      let {listWidthPix} = this.props;
      
      let itemsListWidthPix;
      if(ST.isNumber(listWidthPix) == false)
      {
        listWidthPix = WIDTH;
        itemsListWidthPix = listWidthPix-5;
      }else{
        itemsListWidthPix = listWidthPix-5;
      }
      
      if(this.state.isFiltred == false)
      {

        for(let i = 0; i < this.props.items.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.props.items[i].raw)
          {
            isSelected = true;
          }

          const newItem = <ListItemLayout setScroll = {isSelected} itemsListWidthPix={itemsListWidthPix} isSelected= {isSelected} key ={i} val ={ this.props.items[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }
      }else{ 

        for(let i = 0; i < this.state.filterdItems.length; i++)
        {
          let isSelected = false;
          if(this.props.selectedVal == this.state.filterdItems[i].raw)
          {
            isSelected = true;
          }
          const newItem = <ListItemLayout setScroll = {i == 0?true: false}  isSelected= {isSelected} key ={i} val ={ this.state.filterdItems[i] } onClick = { this.clickItemHandle } />;
          
          items.push(newItem);
        }

      }

      let searchBlock;
      if(!this.props.disableSearch && this.props.items.length > 12)
      {
        searchBlock = <tr>
                          <td>
                            <div><SearchFldLayoutEdit onChangeDelay = { this.changeSearchHandle } /></div>
                          </td>
                        </tr>;
      }
      
      return(
        <div id="ListLayout" ref = {this.genDiv}  style= {{background: 'white', width: listWidthPix,  boxSizing: 'border-box', zIndex:100, position: 'absolute', border: '1px solid gray'}}>
          <table style={{borderSpacing: '2px', width: listWidthPix, borderCollapse: 'separate'}}>
            <tbody>
              {searchBlock}
              <tr>
                <td style={{padding: '0px', margin: '0px'}}>
                  <div id="ListLayout-item" style={{maxHeight: MAXHEIGHT, width: itemsListWidthPix, overflow: 'auto', boxSizing: 'content-box'}} /*onWheel={this.onWheelHandle}*/>
                    {items}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
      </div>);
    }
}
