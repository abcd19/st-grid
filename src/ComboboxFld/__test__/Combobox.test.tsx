import React from 'react';
import { ComboboxFldLayoutEdit, IComboboxFldLayoutEditProps, SearchFldLayoutEdit } from '../'
import { FieldLayoutEdit, ImgButtonLayout, InputLayout } from './../../StringFld';
import { ListLayout } from './../ListLayout';
import { ListItemLayout } from './../ListItemLayout';
import { shallow, mount } from 'enzyme';

const items = [
  { raw: 'raw1', display: 'display1' },
  { raw: 'raw2', display: 'display2' }];

const mockOpenList = {
  currentTarget: {
    getBoundingClientRect: () => { return { left: 100, top: 100 } }
  }
}

/*let list = undefined;
beforeEach(() => {
  list = [
    {raw: 'raw1', display: 'display1'},
    {raw: 'raw2', display: 'display2'}];
});*/
const setUp = (props?: IComboboxFldLayoutEditProps) => {

  const items = [
    { raw: 'raw1', display: 'display1' },
    { raw: 'raw2', display: 'display2' }];

  return shallow(<ComboboxFldLayoutEdit items={items} {...props} />);
}

describe('Combobox', () => {

  let container: HTMLElement;
  HTMLElement.prototype.scrollIntoView = function() {};
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  it('should created successfully', () => {
    const component = mount(<ComboboxFldLayoutEdit items={items} />, { attachTo: container });
    expect(component.render()).toMatchSnapshot();
  });

  test('value prop', () => {
    const component = setUp({ val: items[0], items: items });
    expect(component.find(FieldLayoutEdit).dive().find(InputLayout).get(0).props.val).toBe(items[0].display);
  });


  test('calback onChange', () => {
    const onChange = jest.fn((val) => {
      expect(val).toStrictEqual(items[0]);
    })
    const component = mount(<ComboboxFldLayoutEdit onChange={onChange} items = {items} />);
    expect(component.find(ListLayout)).toHaveLength(0);
    component.find(ImgButtonLayout).find('div').simulate('click', mockOpenList);
    expect(component.find(ListLayout)).toHaveLength(1);
    expect(component.find(ListLayout).find(ListItemLayout)).toHaveLength(items.length);

    component.find(ListItemLayout).first().find('div').simulate('click', { stopPropagation: () => true });
    expect(component.find(ListLayout)).toHaveLength(0);
    expect(onChange).toHaveBeenCalled();
  });


  test('search test', () => {

    const items = [];
    for (let i = 0; i < 30; i++) {
      items.push({ raw: "text" + i, display: "text" + i })
    }
    const component = mount(<ComboboxFldLayoutEdit items = {items} />, { attachTo: container });
    component.find(ImgButtonLayout).find('div').simulate('click', mockOpenList);
    jest.useFakeTimers();
    component.find(SearchFldLayoutEdit).find('input').simulate('change', { target: { value: '1' } });
    jest.runAllTimers();
    jest.advanceTimersByTime(2000);
    //console.dir(component.find(ListLayout).instance().state)
    component.find(ListLayout).update();
    expect(component.find(ListLayout).find(ListItemLayout)).toHaveLength(12)
    jest.clearAllTimers();


    component.find(SearchFldLayoutEdit).find('input').simulate('change', { target: { value: '' } });
    jest.runAllTimers();
    jest.advanceTimersByTime(2000);
    jest.clearAllTimers();
    component.find(ListLayout).update();
    expect(component.find(ListLayout).find(ListItemLayout)).toHaveLength(30)
  });


  test('click searchBtn', () => {
    const items = [];
    for (let i = 0; i < 30; i++) {
      items.push({ raw: "text" + i, display: "text" + i })
    }

    const component = mount(<ComboboxFldLayoutEdit items = {items} />, { attachTo: container });
    component.find(ImgButtonLayout).find('div').simulate('click', mockOpenList);
    component.find(SearchFldLayoutEdit).find('input').simulate('change', { target: { value: '1' } });
    component.find(SearchFldLayoutEdit).find(ImgButtonLayout).find('div').simulate('click');
    expect(component.find(ListLayout).find(ListItemLayout)).toHaveLength(12)
  });


  /*
   
    it('Инициализация без падений', ()=>{
      const wrapper = mount(<ComboboxFldLayoutEdit />);
      expect(wrapper.find('input')).toHaveLength(1);
    });
  
    it('Проверка атрибута val', ()=>{
      let val = {raw: 'raw1'}
      const wrapper = mount(<ComboboxFldLayoutEdit val={val} items={list}/>);
      expect(wrapper.find('input').props().value).toBe('display1');
    });
  
    it('Передача несуществующего значения в прос val', ()=>{
      let val = {raw: 'raw5Unknown'}
      const wrapper = mount(<ComboboxFldLayoutEdit val={val} items={list}/>);
      expect(wrapper.find('input').props().value).toBe('');
    });
  
    it ('Проверка атрибута clearBtnFlag', ()=>{
      const onChange = jest.fn((val) =>val);
      let val = {raw: 'raw1'}
      const wrapper = mount(<ComboboxFldLayoutEdit val={val} onChange={onChange} items={list} clearBtnFlag/>);
      //Выбрали значение
      expect(wrapper.find('input').props().value).toBe('display1');    
      //Есть кнопка очистить
      expect(wrapper.find('div.st_icon_24_clearCellGray')).toHaveLength(1);
      //кликаем по кнопке очистить
      wrapper.find('div.st_icon_24_clearCellGray').simulate('click');
      //колбек вызван 1 раз
      expect(onChange.mock.calls.length).toBe(1);
      //в значении пустая строка
      expect(onChange.mock.results[0].value).toBe('');
      
    });
  
    it('Проверка onChange',()=>{
  
      const onChange = jest.fn((val) =>val);
      const wrapper = mount(<ComboboxFldLayoutEdit onChange={onChange} items={list} />);
      
      //st_icon_24_combobox
      let btn = wrapper.find('div.st_icon_24_combobox');
      expect(btn).toHaveLength(1);
      //открываем список на выбор
      btn.simulate('click');
      let LIL = wrapper.find(ListLayout).find(ListItemLayout);
      //список содержит 2 элемента
      expect(LIL).toHaveLength(2);
      let firstDiv = LIL.first().find('div');
      //выбираем первый 
      firstDiv.simulate('click');
      //проверяем значение
      expect(onChange.mock.calls.length).toBe(1);
      expect(onChange.mock.results[0].value).toEqual({"display":"display1", "raw": "raw1" });
      //cписок закрылся
      expect(wrapper.find(ListLayout)).toHaveLength(0);
    });
  */

})
