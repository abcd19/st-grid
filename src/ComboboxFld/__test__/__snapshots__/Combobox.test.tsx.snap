// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox should created successfully 1`] = `
<table
  class="st-field-table-project st-linear-table"
  style="padding: 0px; margin: 0px; border-spacing: 0px;"
>
  <tbody>
    <tr>
      <td
        align="center"
        style="padding: 0px; margin: 0px;"
      >
        <input
          class="st-core-input-input"
          placeholder=""
          style="white-space: nowrap; line-height: normal; font-kerning: auto; font-family: \\"Tahoma\\", Helvetica, Arial, sans-serif; font-size: 10pt;"
          type="text"
          value=""
        />
      </td>
      <td
        align="center"
        style="padding: 0px; margin: 0px; width: 24px;"
      >
        <div
          class="st_icon_24_combobox"
          style="width: 24px; height: 24px; background-color: transparent;"
          title="button"
        />
      </td>
      <td
        style="padding: 0px; margin: 0px;"
      />
    </tr>
  </tbody>
</table>
`;
